### ISSUE TITLE
HasQueryFilter with navigation property ignores Include calls deeper than 3 layers

### ISSUE THREAD

**Issue by roysanchez** [dotnet/efcore#21714]
When you have an object graph with at least 4 layers (Class A -> B -> C -> D), setting a GlobalQueryFilter for the second class (B) with a navigation property (B.C.IsActive) causes any Include after C to fail.

### Steps to reproduce
This is a small demo console app:
Models:
``` C#
public class Teacher
{
    public int Id { get; set; }
    public string Name { get; set; }
    public bool IsActive { get; set; }
    [ForeignKey(nameof(Subject))]
    public int SubjectId { get; set; }
    public Subject Subject { get; set; }
}

public class Subject
{
    public int Id { get; set; }
    public string Name { get; set; }
    public bool IsActive { get; set; }
    [ForeignKey(nameof(Room))]
    public int RoomId { get; set; }
    public Room Room { get; set; }
}

public class Room
{
    public int Id { get; set; }
    public string Name { get; set; }
    public bool IsActive { get; set; }
    [ForeignKey(nameof(Student))]
    public int StudentId { get; set; }
    public Student Student { get; set; }
}

public class Student
{
    public int Id { get; set; }
    public string Name { get; set; }
    public bool IsActive { get; set; }
}
```
DbContext:
``` C#
public class AppContext : DbContext
{
    public DbSet<Teacher> Teachers { get; set; }
    public DbSet<Subject> Subjects { get; set; }
    public DbSet<Room> Rooms { get; set; }
    public DbSet<Student> Students { get; set; }

    public static readonly ILoggerFactory MyLoggerFactory = LoggerFactory.Create(builder => { builder.AddConsole(); });

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.EnableSensitiveDataLogging();
        optionsBuilder.UseLoggerFactory(MyLoggerFactory);
        optionsBuilder.UseSqlServer(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true;initial catalog=TestDb");
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        base.OnModelCreating(modelBuilder);

        modelBuilder.Entity<Subject>().HasQueryFilter(subject => subject.Room.IsActive);

        modelBuilder.Entity<Student>().HasData(
            new Student { Id = 1, Name = "Student1", IsActive = true }
        );

        modelBuilder.Entity<Room>().HasData(
            new Room { Id = 1, Name = "Room1", IsActive = true, StudentId = 1 }
        );

        modelBuilder.Entity<Subject>().HasData(
            new Subject { Id = 1, Name = "Subject1", IsActive = true, RoomId = 1 }
        );

        modelBuilder.Entity<Teacher>().HasData(
            new Teacher { Id = 1, Name = "Teacher1", IsActive = true, SubjectId = 1 }
        );
    }
}
```

Program:
``` C#
class Program
{
    static void Main(string[] args)
    {
        using (var db = new AppContext())
        {
            db.Database.EnsureDeleted();
            db.Database.EnsureCreated();

            var teacher = db.Teachers.Include(teacher => teacher.Subject.Room.Student)
                .First(t => t.Id == 1);

            if (teacher.Subject.Room.Student is null)
            {
                throw new NullReferenceException();
            }
        }
    }
}
```

Setting the Global Query `HasQueryFilter(subject => subject.Room.IsActive)` disables any Include/ThenInclude call after Room, making the Student property null.

The student entity is not included in the generated query:
``` SQL
SELECT TOP(1)
    [t].[Id], [t].[IsActive], [t].[Name], [t].[SubjectId], [t0].[Id], [t0].[IsActive], [t0].[Name], [t0].[RoomId], [t0].[Id0], [t0].[IsActive0], [t0].[Name0], [t0].[StudentId]
FROM [Teachers] AS [t]
    INNER JOIN (
          SELECT [s].[Id], [s].[IsActive], [s].[Name], [s].[RoomId], [r].[Id] AS [Id0], [r].[IsActive] AS [IsActive0], [r].[Name] AS [Name0], [r].[StudentId]
    FROM [Subjects] AS [s]
        INNER JOIN [Rooms] AS [r] ON [s].[RoomId] = [r].[Id]
    WHERE [r].[IsActive] = CAST(1 AS bit)
      ) AS [t0] ON [t].[SubjectId] = [t0].[Id]
WHERE [t].[Id] = 1
```

### Further technical details

EF Core version: 3.1.6
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1.6
Operating system: Windows 10.0.17763.1339
IDE: Visual Studio 2019 16.6.3

Thanks,

**Comment by maumar** [comment]
I verified that this repros in 3.1.6 and that it works correctly in current bits, closing

**Comment by davethieben** [comment]
> I verified that this repros in 3.1.6 and that it works correctly in current bits, closing

what is meant by "current bits"? does that mean v5? or did it make it in to 3.1.x?

**Comment by smitpatel** [comment]
This is fixed in 5.0.0-rc1
